src/compiler/checker.ts

```diff
@@ -31353,6 +31356,42 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         }
     }
 
+    function checkGrammarRegularExpressionLiteral(node: RegularExpressionLiteral) {
+        const sourceFile = getSourceFileOfNode(node);
+        if (!hasParseDiagnostics(sourceFile)) {
+            let lastError: DiagnosticWithLocation | undefined;
+            const scanner = createScanner(
+                sourceFile.languageVersion,
+                /*skipTrivia*/ true,
+                sourceFile.languageVariant,
+                sourceFile.text,
+                (message, length, arg0) => {
+                    // emulate `parseErrorAtPosition` from parser.ts
+                    const start = scanner.getTokenEnd();
+                    if (message.category === DiagnosticCategory.Message && lastError && start === lastError.start && length === lastError.length) {
+                        const error = createDetachedDiagnostic(sourceFile.fileName, sourceFile.text, start, length, message, arg0);
+                        addRelatedInfo(lastError, error);
+                    }
+                    else if (!lastError || start !== lastError.start) {
+                        lastError = createFileDiagnostic(sourceFile, start, length, message, arg0);
+                        diagnostics.add(lastError);
+                    }
+                },
+                node.pos,
+                node.end - node.pos,
+            );
+            Debug.assert(scanner.scan() === SyntaxKind.SlashToken);
+            Debug.assert(scanner.reScanSlashToken(/*reportErrors*/ true) === SyntaxKind.RegularExpressionLiteral);
+            return !!lastError;
+        }
+        return false;
+    }
+
+    function checkRegularExpressionLiteral(node: RegularExpressionLiteral) {
+        checkGrammarRegularExpressionLiteral(node);

```

Comment:
```suggestion
        addLazyDiagnostic(() => checkGrammarRegularExpressionLiteral(node));

```
you also might want to cache the state of having performed the check, since, as an expression, we might call `check` on it a _lot_ - deferral will remove that cost for LS operations, but a cache will help make it so we only do this rescan once.

Reply:
What's the purpose of `addLazyDiagnostic` here? Does it cache?

Reply:
> you also might want to cache the state of having performed the check, since, as an expression, we might call `check` on it a _lot_

I'm surprised we need an independent cache when we already have `checkExpressionCached`. It feels like we've significantly weakened caching over the years.

Reply:
Are you suggesting to cache *and* to call `addLazyDiagnostic`, or is caching alone sufficient?

Reply:
> What's the purpose of `addLazyDiagnostic` here? Does it cache?

It appears to defer some non-consequential checks until later so that some LS features can use the checker without eagerly checking things that produce diagnostics but also don't affect types.

Reply:
Exactly - `addLazyDiagnostic` is actually eager in the CLI/diagnostics request case and immediately invokes the passed function. In the LS, though, it saves off that work and only does it later when diagnostics actually get requested (or never, if the checker is discarded due to an edit, as is often the case).

> Are you suggesting to cache and to call addLazyDiagnostic, or is caching alone sufficient?

Caching alone alleviates most of my concerns (since `checkExpreessionWithContextualType` is going to check the same expression without caching a bunch), but you'll get slightly better LS perf by deferring it, too.